spec/java/util/HashSet.Spliterator.lsl

```diff
@@ -0,0 +1,221 @@
+            hi = this.fence;
+        }
+        result = hi;
+    }
+
+
+    @AutoInline proc _throwNPE (): void
+    {
+        action THROW_NEW("java.lang.NullPointerException", []);
+    }
+
+
+    proc _checkForComodification (): void
+    {
+        val modCount: int = HashSetAutomaton(this.parent).modCount;
+        if (this.expectedModCount != modCount)
+            action THROW_NEW("java.util.ConcurrentModificationException", []);
+    }
+
+
+    // constructors
+
+    constructor *.HashSet_KeySpliterator (@target self: HashSet_KeySpliterator, source: HashMap, origin: int, fence: int, est: int, expectedModCount: int)
+    {
+        this.index = origin;
+        this.fence = fence;
+        this.est = est;
+        this.expectedModCount = expectedModCount;
+    }
+
+
+    // methods
+
+    fun *.estimateSize (@target self: HashSet_KeySpliterator): long
+    {
+        _getFence();
+        result = this.est as long;
+    }
+
+
+    fun *.characteristics (@target self: HashSet_KeySpliterator): int
+    {
+        action ASSUME(this.parent != null);
+
+        var mask: int = 0;
+        val length: int = HashSetAutomaton(this.parent).length;
+        if (this.fence < 0 || this.est == length)
+            mask = 64;
+
+        result = mask | 1;
+    }
+
+
+    fun *.forEachRemaining (@target self: HashSet_KeySpliterator, userAction: Consumer): void
+    {
+        action ASSUME(this.parent != null);
+
+        if(userAction == null)
+            _throwNPE();
+
+        var hi: int = this.fence;
+        var mc: int = this.expectedModCount;
+        var i: int = this.index;
+        val length: int = HashSetAutomaton(this.parent).length;
+
+        if(hi < 0)
+        {
+            this.expectedModCount = HashSetAutomaton(this.parent).modCount;
+            mc = this.expectedModCount;
+            // problem
+            // How correctly write such string "hi = fence = (tab == null) ? 0 : tab.length;"
+            // As I see this condition "tab == null" we mustn't check, because we don't have "map.table" field;
+            this.fence = length;
+            hi = this.fence;
+        }
+
+        // Original condition: "if (tab != null && tab.length >= hi && (i = index) >= 0 && (i < (index = hi) || current != null))"
+        // This is correct this condition translation ?
+        this.index = hi;
+        if (length > 0 && length >= hi && i >= 0 && i < this.index)
+        {
+            // I must think about this:
+            action LOOP_WHILE(
+                i < hi,
+                forEachRemaining_loop(userAction, i)
+            );
+
+            val modCount: int = HashSetAutomaton(this.parent).modCount;
+            if (modCount != mc)
+                action THROW_NEW("java.util.ConcurrentModificationException", []);
+        }
+    }
+
+
+    @Phantom proc forEachRemaining_loop (userAction: Consumer, i: int): void
+    {
+        val key: Object = action LIST_GET(this.keysStorage, this.index);
```

Comment:
The `userAction` will each an every time receive the same `key` stored at index `this.index`.

Reply:
Fixed